### ISSUE TITLE
"setarch i386 rpmbuild ... " no longer working for building 32-bit packages on x86_64 host

### ISSUE THREAD

**Issue by pghmcfc** [rpm-software-management/rpm#2850]
**Describe the bug**
Since upgrading from Fedora 38 to Fedora 39, I cannot create my Oracle Java 8 (i586) packages on my x86_64 host by running rpmbuild via "setarch i386". If I try to do it, it just does a 64-bit build as if the setarch command hadn't been used.

In fact, on Fedora 39 with rpm 4.19.1 I find that "setarch i386 rpm --showrc" produces 
 exactly the same output as "rpm --showrc". Is that expected?

**To Reproduce**
1. I have javapackages-tools installed
2. setarch i386 rpmbuild -bb java-1.8.0-oracle.spec
3. I get x86_64 packages.

My java spec file can be found here:
https://github.com/pghmcfc/java-1.8.0-oracle/blob/master/java-1.8.0-oracle.spec

**Expected behavior**
I expected to get i586 packages.

**Output**
...
Wrote: /home/paul/cfo-build/java-1.8.0-oracle/trunk/x86_64/java-1.8.0-oracle-plugin-1.8.0.391-1.0.cf.x86_64.rpm
Wrote: /home/paul/cfo-build/java-1.8.0-oracle/trunk/x86_64/java-1.8.0-oracle-1.8.0.391-1.0.cf.x86_64.rpm
Wrote: /home/paul/cfo-build/java-1.8.0-oracle/trunk/x86_64/java-1.8.0-oracle-src-1.8.0.391-1.0.cf.x86_64.rpm
Wrote: /home/paul/cfo-build/java-1.8.0-oracle/trunk/x86_64/java-1.8.0-oracle-devel-1.8.0.391-1.0.cf.x86_64.rpm
Wrote: /home/paul/cfo-build/java-1.8.0-oracle/trunk/x86_64/java-1.8.0-oracle-headless-1.8.0.391-1.0.cf.x86_64.rpm
Wrote: /home/paul/cfo-build/java-1.8.0-oracle/trunk/x86_64/java-1.8.0-oracle-javafx-1.8.0.391-1.0.cf.x86_64.rpm

**Environment**
 - OS / Distribution: Fedora 39
 - Version: rpm 4.19.1

**Additional context**
Output of `rpm --showrc | awk 'NF < 1 { exit } { print }'`:

Fedora 39:
```
ARCHITECTURE AND OS:
build arch            : x86_64
compatible build archs: x86_64_v3 x86_64_v2 x86_64 noarch
build os              : Linux
compatible build os's : Linux
install arch          : x86_64_v3
install os            : Linux
compatible archs      : x86_64_v3 x86_64_v2 x86_64 amd64 em64t athlon noarch i686 i586 i486 i386 fat
compatible os's       : Linux
```
Fedora 39 with `setarch i386`:
```
ARCHITECTURE AND OS:
build arch            : x86_64
compatible build archs: x86_64_v3 x86_64_v2 x86_64 noarch
build os              : Linux
compatible build os's : Linux
install arch          : x86_64_v3
install os            : Linux
compatible archs      : x86_64_v3 x86_64_v2 x86_64 amd64 em64t athlon noarch i686 i586 i486 i386 fat
compatible os's       : Linux
```
I don't have an F-38 system to try any longer, but I have a Rocky Linux 9 VM running rpm 4.16.1 that still works as I expected:

Without `setarch`:
```
ARCHITECTURE AND OS:
build arch            : x86_64
compatible build archs: x86_64 noarch
build os              : Linux
compatible build os's : Linux
install arch          : x86_64
install os            : Linux
compatible archs      : x86_64 amd64 em64t athlon noarch i686 i586 i486 i386 fat
compatible os's       : Linux
```
With:
```
ARCHITECTURE AND OS:
build arch            : i686
compatible build archs: i686 i586 i486 i386 noarch fat
build os              : Linux
compatible build os's : Linux
install arch          : i686
install os            : Linux
compatible archs      : i686 i586 i486 i386 noarch fat
compatible os's       : Linux
```

If this is now expected behaviour, what is the supported way of building for a 32-bit target on x86_64? I can use mock if really necessary but it seems like overkill for a repackaging job like this.

**Comment by dmnks** [comment]
This is caused by the [x86-64 architecture levels patch](https://github.com/rpm-software-management/rpm/commit/cd46c1704ccd8eeb9b600729a0a1c8738b66b847) shipped in 4.19.0. When determining the target architecture internally in `rpmPlatform()`, we use `uname(2)` to obtain the system architecture, however with the above patch, it gets overridden from `i386` (when using `setarch(8)`) to the respective `x86_64` level that's detected.

As a workaround, you can use `rpmbuild --target i386` which actually overrides native architecture detection as expected.

**Comment by dmnks** [comment]
Just FTR, I tried this myself in a Fedora 38 and Fedora 39 container, and it worked. Or rather, the `--target i386` override on Fedora 39 resulted in the same outcome as when using `setarch i386` on Fedora 38, which is that rpmbuild attempted to do a 32bit build. That however failed for me (in both cases) with the following GCC error:

```
[...]
gcc -O2 -g -march=i386 -mtune=i686  -c -o color.o color.c
cc1: error: CPU you selected does not support x86-64 instruction set
[...]
```

**Comment by pghmcfc** [comment]
Using `--target i386` doesn't help me either. An `%ifarch %ix86` clause is not used, and an `%ifarch x86_64` one is.

**Comment by dmnks** [comment]
Yup, having discussed this with the rest of the team, we have to look at fixing this in RPM after all. It definitely is a regression introduced in 4.19 as `setarch` had always worked before. I'm putting it in our queue now.
